<!--
  Launch file to perform mapping using octomap with the radar
 -->
<launch>

  <!-- using best_range_res or best_doppler_res radar config? -->
  <arg name="rng_res" default="false" doc="true: best_range_res cfg; false: best_doppler_res cfg"/>

  <arg name="pcl_limit_filter_topic" default="mmWaveDataHdl/RScan"/>

  <!-- Add nodes to filter/limit Radar data before using it for mapping -->
  <include file="$(find radar_rig)/launch/radar_limit_filters.launch">
     <arg name="radar_pcl" value="$(arg pcl_limit_filter_topic)"/>
  </include>

  <!-- Perform ROR filtering on output from pcl passthrough filters -->
  <node pkg="nodelet" type="nodelet" name="rzi_ROR_filt_mapping" args="standalone pcl/RadiusOutlierRemoval" output="screen">
    <remap from="~input" to="rzi_filt_out"/>
    <remap from="~output" to="rzi_ROR_filter_out"/>
    <!-- best_range_res param values -->
    <param if="$(arg rng_res)" name="min_neighbors" value="4"/>
    <param if="$(arg rng_res)" name="radius_search" value="0.10"/>
    <!-- best_doppler_res param values -->
    <param unless="$(arg rng_res)" name="min_neighbors" value="4"/>
    <param unless="$(arg rng_res)" name="radius_search" value="0.15"/>
  </node>

  <!-- Create "filtered" 3D map from ROR filtered rzi passthrough-filtered pcl -->
  <node pkg="octomap_server" type="octomap_server_node" name="octomap_ROR" output="screen">
    <rosparam command="load" file="$(find radar_rig)/params/octomap.yaml"/>

    <remap from="cloud_in" to="rzi_ROR_filter_out"/>
    <remap from="octomap_point_cloud_centers" to="radar/point_cloud_centers"/>
    <remap from="occupied_cells_vis_array" to="radar/occupied_cells_vis_array"/>
    <remap from="free_cells_vis_array" to="radar/free_cells_vis_array"/>
    <remap from="projected_map" to="radar/map"/>
    <remap from="octomap_binary" to="radar/octomap_binary"/>
    <remap from="octomap_full" to="radar/octomap_full"/>
  </node>

</launch>
